tests/compile_tests.cpp

```diff
@@ -730,3 +730,53 @@ TEST_CASE("test compile change streams") {
   out = cfun({array(1.0f), array(2.0f)})[0];
   CHECK_EQ(out.primitive().stream(), s);
 }
+
+TEST_CASE("test compile lambda") {
+  auto fun = [](const std::vector<array>& inputs) {
+    return std::vector<array>{abs(inputs[0])};
+  };
+
+  auto out = compile(+fun)({array(-1)});
+  CHECK_EQ(out[0].item<int>(), 1);
+
+  decltype(compile(nullptr)) c_local_fun;
+  {
+    auto local_fun = [](const std::vector<array>& inputs) {
+      return std::vector<array>{abs(inputs[0])};
+    };
+    c_local_fun = compile(+local_fun);
```

Comment:
I understand what is going on but I kinda hate that this is allowed. It will be a head scratcher for many people down the road. Not sure it is changeable or even documentable cause it is in the calling code and not in MLX.

The weirdness comes from the fact that a capture-less lambda is convertible to a function pointer (let alone the + operator weirdness) and persists further than the pointer pointing to it.

Reply:
Right so this is a really interesting case (sorry I meant to mention it in the PR comment).

If a lambda is capture-less it is convertible to a function pointer and so we don't need to manage its lifetime (we can treat it just like a plain old function). What I did there was an suboptimal attempt to enable that optimization.

What happens if you don't add the `+` is you get a compiler error which for sure will be confusing. We could document that.. or we could just remove the specialization for now and treat capture-less lambdas the same way we treat captured ones. (I was exploring a third option which used compile time checking if the lambda has a capture and using a specialization if not, but I didn't get it to work yet and it gets pretty messy in the header which I don't love).

Reply:
Yeah exactly. I think treating it as a normal lambda would be ok. It would incur just the cost of allocating a pointer on the heap (that simply points to a function pointer).

It isn't too bad honestly. I just don't think a lot of people will get it the first time and the fact that you have to use the `+` idiom to not get a compiler error is kinda bad.

Reply:
Oh I guess another benefit is that you can "recompile" the same capture-less lambda and it won't recompile cause it will be just a pointer while if we treat it as a lambda it will be recompiled.

Reply:
Right that is the main benefit I was thinking of. But I don't like the `+` thing either so let's pull remove it and we can try to do a better fix later.

Reply:
If the only problem is the `+` then having a pointer type to the function sth like `using ArrayFunPtr = std::vector<array>(*)(const std::vector<array>&);` and then used as

```cpp
auto local_fun = [](const std::vector<array>& inputs) {
  return std::vector<array>{abs(inputs[0])};
};
c_local_fun = compile((ArrayFunPtr)local_fun);
```

is probably more understandable. (didn't compile it but it should work or with `static_cast<ArrayFunPtr>(...)`)